### PR TITLE
5330 category tables

### LINKED ISSUE #5330: Create `category` tables
> Note: updated for category groups

For now:

```mermaid
erDiagram
    item ||--o{ category_item_join : id--item_id
    category ||--o{ category_item_join : id--category_id
    category_group ||--o{ category : id--category_group_id

    category_group {
        text id
        text name
        text root_id
    }

    category {
        text id
        text parent_id
        text description
        text name
        text category_group_id
    }
    category_item_join {
        text category_id
        text item_id
    }
    item {
        text id
    }
```

Categories exist in a hierarchy (e.g. Drug > Analgesic > NSAID) - if an item is an NSAID, CategoryGroup would be Drug, then it should have 2 `item_category_join` entries, one for each level, so item can easily be queried by category.

Future state: 

This design should allow for further extension in future, to include tags (allow us to query at a specific level of the hierarchy, across many branches easily) and categorise different domain types:

```mermaid
erDiagram
    category_group ||--o{ category : id--category_group_id
    item ||--o{ category_item_join : id--item_id
    category ||--o{ category_item_join : id--category_id
    category ||--o{ category_name_join : id--category_id
    name ||--o{ category_name_join : id--name_id
    category_group {
        text id
        text name
        text root_id
        domain_type domain
    }
    category {
        text id
        text parent_id
        text category_group_id
        text description
        text name
        domain_type domain
    }
    category_item_join {
        text category_id
        text item_id
    }
    
    category_item_join {
        text category_id
        text item_id
    }
    name {
        text id
    }
    item {
        text id
    }
    category_name_join {
        text category_id
        text name_id
    }
    domain_type {
        variant item
        variant name
    }
```

### Context

Andrei's original design: 

https://github.com/msupply-foundation/open-msupply/blob/c93ed5cc7ef3763aeda4cfcbd1de2cc49f651c7c/categories%20and%20properties.md

### PR DESCRIPTION
**Pull request by lache-melvin**
Fixes #5330

# 👩🏻‍💻 What does this PR do?

Adds repository layer/migrations for category_group, category and item_category join table.

## 💌 Any notes for the reviewer?

# 🧪 Testing

- [ ] Nothing to test yet - just that migration works correctly

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

**Review by jmbrunskill — COMMENTED** [review]
I think I'm missing a lot of context and discussion on this, but from what i can see, it feels like we should probably collapse `category_group` into the `category` table.

Category group just seems to be the head of the tree?

The design and naming seems to be different to @andreievg's diagram too which is confusing, and I'm not sure the reasons why?

Would be good to get the reasoning/thinking documented maybe a KDD or link to discussion?

File: server/repository/src/migrations/v2_04_01/category_and_item_categories.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    fn migrate(&self, connection: &StorageConnection) -> anyhow::Result<()> {
+        // category_group table
+        sql!(
+            connection,
+            r#"
+            CREATE TABLE category_group (
+                id TEXT PRIMARY KEY NOT NULL,
+                name TEXT NOT NULL,
+                -- no referential constraint due to circular dependency during sync integration
```
[the remark this answers is not in the document]
Comment (lache-melvin) [reply server/repository/src/migrations/v2_04_01/category_and_item_categories.rs:19]: Let's discuss on Monday - but this category groups came about for two reasons:

- The categorisation "type" (still trying to find clear language haha) - e.g. ATC, national drug classification ... it's not really a root category?
- A separate tag to easier query items at a particular level of categorisation, across different branches (that's where category_level came from, but I found that name confusing... hence `group` 😅)

**Comment by lache-melvin** [comment]
Hey @jmbrunskill! I have added a KDD for this, and with the plan to manage categories from OG for now, I think safe to not yet include the `category_group` piece and keep that bit of complexity out until its needed? Let me know what you think 🙏

**Review by jmbrunskill — APPROVED** [review]
Looks great, Thanks Lache!

**Pull request merged**